### PR TITLE
LLVM 18 support

### PR DESCRIPTION
**Pull request by franz**

### CODE REVIEW

**Comment by franz** [comment]
A few tests still fail:

```
The following tests FAILED:
	106 - regression/barrier_between_two_for_loops_cbs (Failed)
	142 - regression/test_alignment_with_dynamic_wg_114_cbs (Failed)
	144 - regression/test_alignment_with_dynamic_wg_117_cbs (Failed)
	146 - regression/test_alignment_with_dynamic_wg_225_cbs (Failed)
	148 - regression/test_alignment_with_dynamic_wg_173_cbs (Failed)
	150 - regression/test_alignment_with_dynamic_wg_183_cbs (Failed)
	152 - regression/test_alignment_with_dynamic_wg_283_cbs (Failed)
	154 - regression/test_alignment_with_dynamic_wg_332_cbs (Failed)
	156 - regression/test_alignment_with_dynamic_wg_323_cbs (Failed)
```

...all of them are related to CBS. @fodinabor  any ideas ? do you have time to take a look ?

File: cmake/LLVM.cmake [hunk 1]
```diff
@@ -788,7 +788,12 @@ if(NOT CLANG_LINK_TEST)
        CompilerInstance CI;
        CompilerInvocation &pocl_build = CI.getInvocation();
 
+       #if (LLVM_MAJOR < 18)
```
Comment (pjaaskel) [inline cmake/LLVM.cmake:791]: Shouldn't cpp macros start from column 0?

**Review by pjaaskel — CHANGES_REQUESTED** [review]
Tiny nit. Can you add a LLVM 18 CI with this PR?

**Comment by pjaaskel** [comment]
Sorry we are a bit busy too, but we can disable CBS (or XFAIL the tests) for LLVM 18 until it's fixed to get the LLVM 18 support in.

File: cmake/LLVM.cmake (hunk 1 above)
Comment (franz) [reply cmake/LLVM.cmake]: fixed

File: CMakeLists.txt [hunk 2]
```diff
@@ -888,9 +888,9 @@ endif()
 # constant addrspace variables, and stack protector likely slows
 # down the kernels, so it needs to be determined whether it's worth
 # the trouble.
-set(DEFAULT_KERNEL_CL_FLAGS  "-xcl -fno-stack-protector -fPIC ${FLOATCONV_FLAG} ${OPAQUE_PTR_FLAGS}")
-set(DEFAULT_KERNEL_C_FLAGS "-xc -std=c11 -D__CBUILD__ -fno-math-errno -fno-stack-protector -fPIC ${FLOATCONV_FLAG} ${OPAQUE_PTR_FLAGS}")
-set(DEFAULT_KERNEL_CXX_FLAGS "-xc++ -std=c++11 -fno-stack-protector -fPIC ${FLOATCONV_FLAG} ${OPAQUE_PTR_FLAGS}")
+set(DEFAULT_KERNEL_CL_FLAGS  "-xcl -fno-stack-protector -fPIC -fno-jump-tables ${FLOATCONV_FLAG} ${OPAQUE_PTR_FLAGS}")
```
Comment (pjaaskel) [inline CMakeLists.txt:891]: Can you add a comment why we add no-jump-tables? It's easy to forget.

File: cmake/LinkTestClang.cc [hunk 3]
```diff
@@ -0,0 +1,62 @@
+#include <clang/Basic/Diagnostic.h>
```
Comment (pjaaskel) [inline cmake/LinkTestClang.cc:1]: Can you add the (c)/license text?

File: cmake/LinkTestLLVM.cc [hunk 4]
```diff
@@ -0,0 +1,22 @@
+#include "llvm/IR/LLVMContext.h"
```
Comment (pjaaskel) [inline cmake/LinkTestLLVM.cc:1]: Same here. License text is missing.

File: cmake/LinkTestLLVM.cc [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+#include "llvm/IR/Module.h"
```
Comment (pjaaskel) [inline cmake/LinkTestLLVM.cc:2]: Please use #include <> as we are not in LLVM source tree.

File: lib/CL/pocl_llvm_build.cc [hunk 6]
```diff
@@ -404,6 +404,7 @@ int pocl_llvm_build_program(cl_program program,
   }
 
   ss << "-fno-builtin ";
+  ss << "-fno-jump-tables ";
```
Comment (pjaaskel) [inline lib/CL/pocl_llvm_build.cc:407]: Also pls add the comment here so we won't forget (I already did).

File: lib/llvmopencl/DebugHelpers.cc [hunk 7]
```diff
@@ -34,6 +34,7 @@ IGNORE_COMPILER_WARNING("-Wunused-parameter")
 #include "llvm/IR/Instructions.h"
 #include "llvm/IR/Module.h"
 #include "llvm/Transforms/Utils/BasicBlockUtils.h"
+#include "llvm/Analysis/RegionInfo.h"
```
Comment (pjaaskel) [inline lib/llvmopencl/DebugHelpers.cc:37]: Pls use <>

File: lib/llvmopencl/DebugHelpers.cc [hunk 8]
```diff
@@ -154,39 +157,50 @@ void dumpCFG(
...
-      s << "\tsubgraph cluster" << pr->GetID() << " {" << std::endl;
-
-      for (ParallelRegion::iterator i = pr->begin(), e = pr->end(); 
-           i != e; ++i) {
-        BasicBlock *b = *i;
-        printBasicBlock(
-          b, s, highlights != NULL && 
-          highlights->find(b) != highlights->end());
-        regionBBs.insert(b);
+  if (Regions!=nullptr && Regions->size()) {
```
Comment (pjaaskel) [inline lib/llvmopencl/DebugHelpers.cc:160]: White space issue, can you run the formatter?

File: lib/llvmopencl/LLVMUtils.h [hunk 9]
```diff
@@ -92,6 +96,17 @@ bool isKernelToProcess(const llvm::Function &F);
 
 llvm::Metadata *createConstantIntMD(llvm::LLVMContext &C, int32_t Val);
 
+// fixes switch statements that have a default case that is a simple unreachable
```
Comment (pjaaskel) [inline lib/llvmopencl/LLVMUtils.h:99]:
```suggestion
// Fixes switch statements that have a default case that is a simple unreachable
```

File: lib/llvmopencl/LLVMUtils.h [hunk 10]
```diff
@@ -92,6 +96,17 @@ bool isKernelToProcess(const llvm::Function &F);
...
+// instruction. LLVM does this as "optimization", however it breaks the
+// (post) dominator-tree analysis, because the unreachable instruction creates
+// an additional function exit path. PoCL's ImplicitConditionalBarriers pass
+// then erroneously adds barriers because it thinks some basic blocks are
+// inside conditional blocks.
```
Comment (pjaaskel) [inline lib/llvmopencl/LLVMUtils.h:104]:
```suggestion
// inside conditional blocks although they are not.
```

File: lib/llvmopencl/LLVMUtils.h [hunk 11]
```diff
@@ -92,6 +96,17 @@ bool isKernelToProcess(const llvm::Function &F);
...
+// TODO this is fragile, LLVM can introduce more optimizations that create
+// unreachable blocks. However i couldn't find any working way to make PDT
```
Comment (pjaaskel) [inline lib/llvmopencl/LLVMUtils.h:106]:
```suggestion
// unreachable blocks. However I couldn't find any working way to make PDT
```

File: lib/llvmopencl/LLVMUtils.h [hunk 12]
```diff
@@ -92,6 +96,17 @@ bool isKernelToProcess(const llvm::Function &F);
...
+// ignore blocks with unreachable inst.
```
Comment (pjaaskel) [inline lib/llvmopencl/LLVMUtils.h:107]:
```suggestion
// ignore blocks with an unreachable inst.
```

File: lib/llvmopencl/OptimizeWorkItemGVars.cc [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+// into a Phi followed by a single load:
+//   %p = phi [ @_local_id_x, %sw.branch.0 ], [ @_local_id_y, %sw.branch.1 ]...
+//   load i64, ptr %p
+//
+// the Phi is then replaced by the phis2allocas pass with an alloca + stores &
+// loads. Since the workgroup pass can only deal with loads from the special WI
+// variables, this ends up leaving unresolved symbols in the final binary.
+//
+//
+// TODO: we should replace the magic global variables with function calls
```
Comment (pjaaskel) [inline lib/llvmopencl/OptimizeWorkItemGVars.cc:17]:
```suggestion
// TODO: we should replace the magic global variables marking the work-item IDs with function calls or intrinsics.
```

File: lib/llvmopencl/OptimizeWorkItemGVars.cc [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+      }
+      if (LI->getFunction() == &F) {
+        GVUsers.push_back(LI);
+      }
+    }
+
+    if (GVUsers.size() > 1) {
+      Changed = true;
+      IRBuilder<> Builder(&*(F.getEntryBlock().getFirstInsertionPt()));
+/*
```
Comment (pjaaskel) [inline lib/llvmopencl/OptimizeWorkItemGVars.cc:93]: Please remove commented out code.

File: lib/llvmopencl/OptimizeWorkItemGVars.cc [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+      Builder.CreateStore(FirstLoad, ReplAlloca);
+      LoadInst *SecondLoad = Builder.CreateLoad(
+          GVar->getValueType(),
+          ReplAlloca, Twine(GVarName, "_load"));
+*/
+      LoadInst *ReplLoad = Builder.CreateLoad(
+          GVar->getValueType(),
+          GVar, Twine(GVarName, "_load"));
+    for (auto U : GVUsers) {
+//        U->getType()->dump();
```
Comment (pjaaskel) [inline lib/llvmopencl/OptimizeWorkItemGVars.cc:108]: Please remove commented out code.

File: lib/llvmopencl/WorkitemReplication.cc [hunk 16]
```diff
@@ -127,7 +127,8 @@ bool WorkitemReplicationImpl::processFunction(Function &F) {
 
   parallel_regions[0] = *OriginalParallelRegions;
 
-  //pocl::dumpCFG(F, F.getName().str() + ".before_repl.dot", original_parallel_regions);
+  //pocl::dumpCFG(F, F.getName().str() + ".before_repl.dot",
```
Comment (pjaaskel) [inline lib/llvmopencl/WorkitemReplication.cc:130]: Pls move these to the #if 0 block below?

**Comment by franz** [comment]
```
[2024-02-27 14:50:58.734061576] POCL: in fn read at line 193:
  *** ERROR ***  |     ERROR |  Read error on reentrant_read(fd, req, request->req_size, &request->req_read), fd=9, reason: Bad address
[2024-02-27 14:50:58.734090907] POCL: in fn connectPeer at line 150:
  *** ERROR ***  |     ERROR |  PH: error on read_full(PeerFd, &Reply, sizeof(Reply), netstat): Connection reset by peer
```

@jansol  is the RDMA CI working nowadays ?

**Comment by jansol** [comment]
#1433 should fix the RDMA CI

**Comment by franz** [comment]
Hi @fodinabor, do you have time to look into the CBS failure a bit more ? Unfortunately removing the instcombine between `loop-barriers` and `barrier-tail` passes still makes `loopvec` fail (because the `phi`s are not handled), but it's unclear to me what causes `instcombine` with CBS to replace the `for.body` with poison.

**Pull request merged**